### PR TITLE
OF-2524: Improving speed of loading MUC history from database

### PR DESCRIPTION
**Pull request by guusdk**
Two changes (see individual commit messages for details):
- add all messages to the history in one batch, to prevent repeated acquiring and releasing of a (cluster-wide) lock
- add database index to speed up query to retrieve messages from the database.

### CODE REVIEW

File: xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MUCPersistenceManager.java [hunk 1]
```diff
@@ -796,23 +797,29 @@ public static void loadHistory(@Nonnull final MUCRoom room, final int maxNumber)
...
                     Date sentDate = new Date(Long.parseLong(rs.getString("logTime").trim()));
                     String subject = rs.getString("subject");
                     String body = rs.getString("body");
                     String stanza = rs.getString("stanza");
-                    room.getRoomHistory().addOldMessage(senderJID, nickname, sentDate, subject, body, stanza);
+                    oldMessages.add(room.getRoomHistory().parseHistoricMessage(senderJID, nickname, sentDate, subject, body, stanza));
+                }
+
+                if (!oldMessages.isEmpty()) {
+                    room.getRoomHistory().addOldMessages(oldMessages);
```
Comment (Fishbowler) [inline xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MUCPersistenceManager.java:812]: Any memory exhaustion concerns here? 
For a short while here, there's double the committed memory.
You'll have seen some esoteric setups, as well as Ignite, which has significant history. You might judge this to be low risk.

Comment (guusdk) [reply xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MUCPersistenceManager.java:812]: The data is limited to that of one room. I don't see how that would be that much of a problem.

Comment (Fishbowler) [reply xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/MUCPersistenceManager.java:812]: True. 100K msgs at 100b per msg would only be a temporary 9MB spike.

**Review by Fishbowler — APPROVED** [review]

**Comment by guusdk** [comment]
cherry-picked to 4.7.

**Pull request merged**